src/server/services/wms/qgswmsrenderer.cpp

```diff
@@ -2637,5 +2730,499 @@ namespace QgsWms
+      {
+        QgsMessageLog::logMessage( errorMsg, "Server", QgsMessageLog::INFO );
+        continue;
+      }
+
+      // build url for vector layer
+      QString typeName = QgsWkbTypes::geometryDisplayString( g.type() );
+      QString url = typeName + "?crs=" + crs;
+      if ( ! param.mLabel.isEmpty() )
+      {
+        url += "&field=label:string";
+      }
+
+      // create vector layer
+      std::unique_ptr<QgsVectorLayer> layer;
+      layer.reset( new QgsVectorLayer( url, param.mName, "memory" ) );
+      if ( !layer->isValid() )
+      {
+        continue;
+      }
+
+      // create feature with label if necessary
+      QgsFeature fet( layer->pendingFields() );
+      if ( ! param.mLabel.isEmpty() )
+      {
+        fet.setAttribute( 0, param.mLabel );
+
+        // init labeling engine
+        QgsPalLayerSettings palSettings;
+        palSettings.readFromLayer( layer.get() );
+        palSettings.enabled = true;
+        palSettings.fieldName = "label"; // defined in url
+        palSettings.priority = 10; // always drawn
+        palSettings.displayAll = true;
+
+        QgsPalLayerSettings::Placement placement = QgsPalLayerSettings::AroundPoint;
+        switch ( g.type() )
+        {
+          case QgsWkbTypes::PointGeometry:
+          {
+            placement = QgsPalLayerSettings::AroundPoint;
+            palSettings.dist = 2; // in mm
+            palSettings.placementFlags = 0;
+            break;
+          }
+          case QgsWkbTypes::PolygonGeometry:
+          {
+            QgsGeometry point = g.pointOnSurface();
+            QgsPoint pt = point.asPoint();
+            placement = QgsPalLayerSettings::AroundPoint;
+
+            QgsPalLayerSettings::Property pX = QgsPalLayerSettings::PositionX;
+            QVariant x( pt.x() );
+            palSettings.dataDefinedProperties().setProperty( pX, x );
+
+            QgsPalLayerSettings::Property pY = QgsPalLayerSettings::PositionY;
+            QVariant y( pt.y() );
+            palSettings.dataDefinedProperties().setProperty( pY, y );
+
+            QgsPalLayerSettings::Property pHali = QgsPalLayerSettings::Hali;
+            QVariant hali( "Center" );
+            palSettings.dataDefinedProperties().setProperty( pHali, hali );
+
+            QgsPalLayerSettings::Property pVali = QgsPalLayerSettings::Vali;
+            QVariant vali( "Half" );
+            palSettings.dataDefinedProperties().setProperty( pVali, vali );
+            break;
+          }
+          default:
+          {
+            placement = QgsPalLayerSettings::Line;
+            palSettings.dist = 2; // in mm
+            palSettings.placementFlags = 10;
+            break;
+          }
+        }
+        palSettings.placement = placement;
+
+        if ( param.mColor.isValid() )
+        {
+          QgsPalLayerSettings::Property pColor = QgsPalLayerSettings::Color;
+          QVariant color( param.mColor );
+          palSettings.dataDefinedProperties().setProperty( pColor, color );
```

Comment:
Why does this use data defined properties? It'd be faster to just set the label color/etc directly